lib/topological_inventory/orchestrator/metric_scaler/prometheus_watcher.rb

```diff
@@ -0,0 +1,105 @@
+require "topological_inventory/orchestrator/metric_scaler/watcher"
+
+module TopologicalInventory
+  module Orchestrator
+    class MetricScaler
+      class PrometheusWatcher < Watcher
+        METRICS_CHECK_INTERVAL = 150
+        METRICS_RANGE          = "10m"
+        METRICS_STEP           = "1m"
+
+        def initialize(deployment_config, deployment_config_name, prometheus_host, logger)
+          super(deployment_config, deployment_config_name, logger)
+          @prometheus_host = prometheus_host
+          @scaling_allowed.value = false
+        end
+
+        def configured?
+          @query && @max_replicas && @min_replicas && @target_usage && @scale_threshold
+        end
+
+        def configure
+          @max_replicas        = deployment_config.metadata.annotations["metric_scaler_max_replicas"]&.to_i        # i.e. "5"
+          @min_replicas        = deployment_config.metadata.annotations["metric_scaler_min_replicas"]&.to_i        # i.e. "1"
+          @target_usage        = deployment_config.metadata.annotations["metric_scaler_target_usage_pct"]&.to_f    # i.e. "50"
+          @scale_threshold     = deployment_config.metadata.annotations["metric_scaler_scale_threshold_pct"]&.to_f # i.e. "20"
+          @query               = deployment_config.metadata.annotations["metric_scaler_prometheus_query"]          # i.e. any query that outputs a single number showing load. Preferably at a 1 minute resolution.
+        end
+
+        def start
+          @thread = Thread.new do
+            logger.info("Watcher thread for #{deployment_config_name} starting")
+            until finished?
+              new_metrics = download_metrics
+              if new_metrics.present?
+                new_metrics.each do |time_value_pair|
+                  metrics << time_value_pair[1].to_f
+                end
+
+                if metrics.average > @scale_threshold
+                  @scaling_allowed.value = true
+                end
+              end
+
+              sleep METRICS_CHECK_INTERVAL
+            end
+            logger.info("Watcher thread for #{deployment_config_name} stopping")
+          end
+        end
+
+        def download_metrics
+          full_url = File.join(prometheus_url, promql_query)
+          response = RestClient.get(full_url, {:accept => :json})
+
+          body = JSON.parse(response.body)
+          raise response.body unless body['status'] == 'success'
+
+          metrics = body.dig('data', 'result').to_a.first
+          if metrics.blank?
+            logger.warn("PrometheusWatcher: Empty result from Prometheus received")
+          end
+          metrics.to_a
+        rescue RestClient::BadRequest => e
+          logger.error("PrometheusWatcher: Bad request: #{e.response.body}")
```

Comment:
So, if the metrics can't get downloaded, that seems like a major issue.  Should we have more than the logging here?

Reply:
Bad request could mean that the PromQL query is wrong in the e2e-deploy template's annotation.
@syncrou what exactly do you mean by having more than logging, Alert? 

In other cases (not HTTP 400) Prometheus is either unavailable or is available, returns HTTP 200, but all values are zero. We can skip scaling in this case but except logging, we will be not able to do more with it

Reply:
@slemrmartin @lindgrenj6 - My thought was, if we have no metrics we have no autoscaling, so it seems potentially a Slack alert is warranted.

Reply:
Yeah thats a good point - how about I follow up with that in a different PR since it entails adding prometheus metrics to the metric_scaler itself, and from there creating an alert on it.

Reply:
@lindgrenj6 that works for me.

Reply:
Will follow up here: https://projects.engineering.redhat.com/browse/TPINVTRY-968